test/topology_custom/test_major_compaction.py

```diff
@@ -159,38 +158,41 @@ async def test_shutdown_drain_during_compaction(manager: ManagerClient):
-    await log.wait_for("update_history_wait: waiting", mark, 30)
-
-    mark = await log.mark()
-    # Start server shutdown
-    logger.info("Shutdown server")
-    stop_task = asyncio.create_task(manager.server_stop_gracefully(server.server_id))
-    # wait until the shutdown drain request is sent to compaction_manager
-    await log.wait_for("Asked to drain", mark, 30)
-    # now resume compaction and let shutdown complete
-    await injection_handler.message()
-    # wait server to shutdown
-    await stop_task
-    # During shutdown, errors mentioning 'seastar::abort_requested_exception' is expected as we do abort the compaction midway.
-    # Verify that the shutdown completed without any other unexpected errors
-    assert len(await log.grep(expr="ERROR .*", filter_expr=".* seastar::abort_requested_exception \(abort requested\)", from_mark=mark)) == 0
+    async with new_test_keyspace(manager, "WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}") as ks:
+        await cql.run_async(f"CREATE TABLE {ks}.{cf} (pk int PRIMARY KEY);")
+        await disable_autocompaction_across_keyspaces(manager, server.ip_addr, ks)
+
+        logger.info("Populating table")
+        await asyncio.gather(*[cql.run_async(f"INSERT INTO {ks}.{cf} (pk) VALUES ({k});") for k in range(100)])
+        await manager.api.keyspace_flush(server.ip_addr, ks, cf)
+
+        # inject error to make compaction wait just before it updates the compaction_history table
+        injection = "update_history_wait"
+        injection_handler = await inject_error_one_shot(manager.api, server.ip_addr, injection)
+
+        log = await manager.server_open_log(server.server_id)
+        mark = await log.mark()
+        # start compaction and wait for it to pause at the injection point
+        logger.info("Start compaction")
+        compaction_task = asyncio.create_task(manager.api.keyspace_compaction(server.ip_addr, ks, cf))
+        await log.wait_for("update_history_wait: waiting", mark, 30)
+
+        mark = await log.mark()
+        # Start server shutdown
+        logger.info("Shutdown server")
+        stop_task = asyncio.create_task(manager.server_stop_gracefully(server.server_id))
+        # wait until the shutdown drain request is sent to compaction_manager
+        await log.wait_for("Asked to drain", mark, 30)
+        # now resume compaction and let shutdown complete
+        await injection_handler.message()
+        # wait server to shutdown
+        await stop_task
+        # During shutdown, errors mentioning 'seastar::abort_requested_exception' is expected as we do abort the compaction midway.
+        # Verify that the shutdown completed without any other unexpected errors
+        assert len(await log.grep(expr="ERROR .*", filter_expr=".* seastar::abort_requested_exception \(abort requested\)", from_mark=mark)) == 0
+
+        # For dropping the keyspace
+        await manager.server_start(server.server_id)
```

Comment:
It should be possible to drop without all servers UP.
Also why is the reconnect necessary?

Anyway, what I don't understand in this entire thing, why are you not using `create_new_test_keyspace` which you introduced at the beginning of the series, it would let you avoid having to indent everything, and the change would be much smaller in scope/behavior (why do we always have to drop the keyspace at the end now, if it was not necessary at the beginning).

Reply:
"async with" it is syntax sugar that helps avoid try/finally indent everywhere.